### PR TITLE
Safe + whitelisting

### PR DESCRIPTION
**Pull request by bausano**
Implementation of a proposal mentioned in this comment: https://github.com/MystenLabs/sui/pull/4887#issuecomment-1285537383

### CODE REVIEW

**Comment by nmboavida** [comment]
> One implication of logical ownership:
> 
> In the case of embedded NFTs, they are currently being preminted to the Launchpad and therefore they initially exist in a state where they do not have logical owner until they get bought.
> 
> One solution is to change to flow as follows:
> 
> 1. Instead of NFT creators preminting the NFT to the launchpad, they will instead only create the `Data` object.
> 2. When the buyers purchase an NFT certificate, the launchpad will mint the NFT object initially as loose and then will call the function `nft::join_nft_data` to transform the loose NFT into an embedded NFT

In fact this means that the interaction surface can be significantly simplified as the embedded and loose NFTs will have pretty much the same flow from an interface standpoint

**Comment by bausano** [comment]
TODO:
- emit events
- changelog
- docs
- some listing strategies such as blacklist or free-for-all

File: sources/safe/safe.move [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    use sui::vec_map::{Self, VecMap};
+    use std::type_name::{Self, TypeName};
+
+    struct Safe has key {
+        id: UID,
+        /// Accounting for deposited NFTs. Each NFT in the object bag is
+        /// represented in this map.
+        refs: VecMap<ID, NftRef>,
+        /// Holds the actual NFT objects as child objects.
+        nfts: ObjectBag,
```
Comment (Suficio) [inline sources/safe/safe.move:21]: Should this still remain an `ObjectBag`? 
Something like a Table would be a bit simpler for the reasons we went with `Bag` for domains.

**Review by nmboavida — APPROVED** [review]
Fast-tracking to merge all to develop, let us add TODOs so we can tackle stuff in the next iterations

File: sources/safe/safe.move (hunk 1 above)
Comment (bausano) [reply sources/safe/safe.move:21]: https://github.com/Origin-Byte/nft-protocol/issues/94

**Pull request merged**